src/main/java/uk/nhs/hee/trainee/details/service/NtnGenerator.java

```diff
@@ -0,0 +1,303 @@
+   *
+   * @param personalDetails The personal details to get the reference number for.
+   * @return The GMC/GDC number, depending on which is valid.
+   */
+  private String getReferenceNumber(PersonalDetails personalDetails) {
+    String gmcNumber = personalDetails.getGmcNumber();
+    return gmcNumber.matches("\\d{7}") ? gmcNumber : personalDetails.getGdcNumber();
+  }
+
+  /**
+   * Get the NTN suffix for the given programme membership.
+   *
+   * @param programmeMembership The programme membership to get the suffix for.
+   * @return The calculated suffix for the programme membership's NTN.
+   */
+  private String getSuffix(ProgrammeMembership programmeMembership) {
+    log.info("Calculating suffix.");
+    String trainingPathway = programmeMembership.getTrainingPathway();
+    log.info("Using training pathway '{}' to calculate suffix.", trainingPathway);
+
+    String suffix = switch (trainingPathway) {
+      case "CCT" -> "C";
+      case "CESR" -> "CP";
+      default -> {
+        List<Curriculum> curricula = programmeMembership.getCurricula();
+        List<Curriculum> sortedCurricula = filterAndSortCurricula(curricula);
+        String firstSpecialtyCode = sortedCurricula.get(0).getCurriculumSpecialtyCode();
+        log.info("Using specialty code '{}' to calculate suffix.", trainingPathway);
+
+        yield Objects.equals(firstSpecialtyCode, "ACA") ? "C" : "D";
+      }
+    };
+
+    log.info("Calculated suffix: '{}'.", suffix);
+    return suffix;
+  }
+
+  /**
+   * Filter the given curricula to only those that are current and sort them alphanumerically.
+   *
+   * @param curricula The curricula to filter and sort.
+   * @return The current curricula, sorted alphanumerically by subtype and code.
+   */
+  private List<Curriculum> filterAndSortCurricula(List<Curriculum> curricula) {
+    LocalDate now = LocalDate.now();
+
+    return curricula.stream()
+        .filter(c ->
+            c.getCurriculumSpecialtyCode() != null && !c.getCurriculumSpecialtyCode().isBlank())
+        .filter(c -> c.getCurriculumStartDate().isBefore(now))
+        .filter(c -> c.getCurriculumEndDate().isAfter(now))
+        .sorted(Comparator
+            .comparing(Curriculum::getCurriculumSubType)
+            .reversed()
+            .thenComparing(Curriculum::getCurriculumSpecialtyCode)
+            .reversed()
+        )
+        .toList();
+  }
+
+  /**
+   * Check whether the given personal details excludes NTN generation.
+   *
+   * @param personalDetails The personal details to check.
+   * @return true if NTN generated cannot continue, else false.
+   */
+  private boolean isExcluded(PersonalDetails personalDetails) {
+    if (personalDetails == null) {
+      log.info("Skipping NTN population as personal details not available.");
+      return true;
+    }
+
+    String gmcNumber = personalDetails.getGmcNumber();
+
+    if (gmcNumber == null || !gmcNumber.matches("\\d{7}")) {
```

Comment:
So this filters out GMC numbers like '95/0830' which I believe are temporary ones.

Reply:
I don't know whether it was a specific decision to do that, but that's the current behaviour in the SQL

Reply:
Am I looking at the right SQL? All I can see is a constraint on null or empty-string GMC numbers, and then another section with some commented-out criteria, viz.:
```
where --Filter to select current or future ProgrammeMemberships depending on future business process
	--NTNComp.GMCNumber is not null
	--and
	--NTNComp.GMCNumber <> 'Unknown'
	--and
	--NTNComp.GMCNumber not like '%N%A%'
	--and
	--NTNComp.GMCNumber <>''
```

Reply:
There were a few different versions which caused confusion. The latest is on the ticket, with the old ones removed.